File: Solid_Principles/Single_Responsibility_Principle/01_SRP.py
# TITLE: Single Responsibility Principle (SRP) - Invoice Example

class Marker:
    def __init__(self, name, color, year, price):
        self.name = name
        self.color = color
        self.year = year
        self.price = price

# ========================================
# THE PROBLEM (VIOLATION):
# Invoice class has 3 responsibilities!
# ========================================

class InvoiceBad:
    def __init__(self, marker, quantity):
        self.marker = marker
        self.quantity = quantity
    
    # Responsibility 1: Calculate total
    def calculate_total(self):
        return self.marker.price * self.quantity
    
    # Responsibility 2: Print invoice
    def print_invoice(self):
        print(f"Invoice: {self.marker.name} x{self.quantity} = ${self.calculate_total()}")
    
    # Responsibility 3: Save to database
    def save_to_db(self):
        print("Saving invoice to database...")

# ========================================
# THE SOLUTION:
# Separate responsibilities
# ========================================

class Invoice:
    def __init__(self, marker, quantity):
        self.marker = marker
        self.quantity = quantity
    
    def calculate_total(self):
        return self.marker.price * self.quantity

class InvoicePrinter:
    def print(self, invoice):
        print(f"Invoice: {invoice.marker.name} x{invoice.quantity} = ${invoice.calculate_total()}")

class InvoiceDao:
    def save_to_db(self, invoice):
        print("Saving invoice to database...")

if __name__ == "__main__":
    marker = Marker("Sharpie", "Blue", 2023, 2.5)
    
    print("=== PROBLEM (SRP Violation) ===")
    bad_invoice = InvoiceBad(marker, 10)
    bad_invoice.print_invoice()
    bad_invoice.save_to_db()
    
    print("\n=== SOLUTION (SRP Compliant) ===")
    invoice = Invoice(marker, 10)
    
    printer = InvoicePrinter()
    printer.print(invoice)
    
    dao = InvoiceDao()
    dao.save_to_db(invoice)

File: Solid_Principles/Single_Responsibility_Principle/01_SRP.js
// TITLE: Single Responsibility Principle (SRP) - Invoice Example

class Marker {
    constructor(name, color, year, price) {
        this.name = name;
        this.color = color;
        this.year = year;
        this.price = price;
    }
}

// ========================================
// THE PROBLEM (VIOLATION):
// Invoice class has 3 responsibilities!
// ========================================

class InvoiceBad {
    constructor(marker, quantity) {
        this.marker = marker;
        this.quantity = quantity;
    }

    // Responsibility 1: Calculate total
    calculateTotal() {
        return this.marker.price * this.quantity;
    }

    // Responsibility 2: Print invoice
    printInvoice() {
        console.log(`Invoice: ${this.marker.name} x${this.quantity} = $${this.calculateTotal()}`);
    }

    // Responsibility 3: Save to database
    saveToDB() {
        console.log("Saving invoice to database...");
    }
}

// ========================================
// THE SOLUTION:
// Separate responsibilities
// ========================================

class Invoice {
    constructor(marker, quantity) {
        this.marker = marker;
        this.quantity = quantity;
    }

    calculateTotal() {
        return this.marker.price * this.quantity;
    }
}

class InvoicePrinter {
    print(invoice) {
        console.log(`Invoice: ${invoice.marker.name} x${invoice.quantity} = $${invoice.calculateTotal()}`);
    }
}

class InvoiceDao {
    saveToDB(invoice) {
        console.log("Saving invoice to database...");
    }
}

// Usage
const marker = new Marker("Sharpie", "Blue", 2023, 2.5);

console.log("=== PROBLEM (SRP Violation) ===");
const badInvoice = new InvoiceBad(marker, 10);
badInvoice.printInvoice();
badInvoice.saveToDB();

console.log("\n=== SOLUTION (SRP Compliant) ===");
const invoice = new Invoice(marker, 10);

const printer = new InvoicePrinter();
printer.print(invoice);

const dao = new InvoiceDao();
dao.saveToDB(invoice);
